src/DotNetty.Transport/Channels/Embedded/EmbeddedChannel.cs

```diff
@@ -0,0 +1,421 @@
+            {
+                if (m == null)
+                    break;
+                futures.Add(this.WriteAsync(m));
+            }
+
+            this.Flush();
+
+            int size = futures.Count;
+            for (int i = 0; i < size; i++)
+            {
+                Task future = futures[i];
+                Debug.Assert(future.IsCompleted);
+                if (future.Exception != null)
+                {
+                    this.RecordException(future.Exception);
+                }
+            }
+
+            this.RunPendingTasks();
+            this.CheckException();
+            return IsNotEmpty(this.outboundMessages);
+        }
+
+
+        private void RecordException(Exception cause)
+        {
+            if (this.lastException == null)
+            {
+                this.lastException = cause;
+            }
+            else
+            {
+                //TODO:
+                // logger.warn(
+                // "More than one exception was raised. " +
+                //        "Will report only the first one and log others.", cause);
+            }
+        }
+
+        /// <summary>
+        /// Mark this <see cref="IChannel"/> as finished. Any further try to write data to it will fail.
+        /// </summary>
+        /// <returns>bufferReadable returns <c>true</c></returns>
+        public bool Finish()
+        {
+            this.CloseAsync();
+            this.CheckException();
+            return IsNotEmpty(this.inboundMessages) || IsNotEmpty(this.outboundMessages);
+        }
+
+        public override Task CloseAsync()
+        {
+            Task future = base.CloseAsync();
+            this.FinishPendingTasks();
+            return future;
+        }
+
+        public override Task DisconnectAsync()
+        {
+            Task future = base.DisconnectAsync();
+            this.FinishPendingTasks();
+            return future;
+        }
+
+        /// <summary>
+        /// Check to see if there was any <see cref="Exception"/> and rethrow if so.
+        /// </summary>
+        public void CheckException()
+        {
+            Exception e = this.lastException;
+            if (e == null)
+                return;
+
+            this.lastException = null;
+            throw e;
+        }
+
+        /// <summary>
+        /// Ensure the <see cref="IChannel"/> is open and if not throw an exception.
+        /// </summary>
+        protected void EnsureOpen()
+        {
+            if (!this.Open)
+            {
+                this.RecordException(new ClosedChannelException());
+                this.CheckException();
+            }
+        }
+
+        private static bool IsNotEmpty(Queue<object> queue)
+        {
+            return queue != null && queue.Count > 0;
+        }
+
+        private static object Poll(Queue<object> queue)
+        {
+            return IsNotEmpty(queue) ? queue.Dequeue() : null;
+        }
+
+        internal class DefaultUnsafe : AbstractUnsafe
```

Comment:
Changed it back to `private` - at the time I ported it wasn't sure if it'd need to be exposed for testing.